### PR TITLE
Implement Image.clone for CanvasKit

### PR DESCRIPTION
**Pull request by dnfield**
## Description

Implements `Image.dispose` and `Image.clone` properly so that https://github.com/flutter/flutter/pull/67100 can be relanded.

Creates a `RefCountedSkiaObjectBox` subclass of `SkiaObjectBox` to support this.

## Related Issues

https://github.com/flutter/flutter/pull/67100
https://github.com/flutter/flutter/pull/66688
https://github.com/flutter/flutter/issues/56482
https://github.com/flutter/engine/pull/21057

## Tests

I added the following tests:

Tests that `Image.dispose` and `Image.clone`/`Image.isCloneOf` work properly.

Tests for `clone` and stacktrace functionality on `RefCountedSkiaObjectBox`.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/image.dart [hunk 1]
```diff
@@ -43,8 +43,14 @@ class CkAnimatedImage implements ui.Image {
   // being garbage-collected, or by an explicit call to [delete].
   late final SkiaObjectBox box;
 
-  CkAnimatedImage(this._skAnimatedImage) {
-    box = SkiaObjectBox(this, _skAnimatedImage as SkDeletable);
+  CkAnimatedImage(SkAnimatedImage skAnimatedImage) : this._(skAnimatedImage, null);
+
+  CkAnimatedImage._(this._skAnimatedImage, SkiaObjectBox? boxToClone) {
+    if (boxToClone != null) {
+      box = boxToClone.clone(this);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/image.dart:50]: Should we assert that the image in `boxToClone` and `_skAnimatedImage` are the same?

Alternatively, we could split this into two specialized constructors.

File: lib/web_ui/lib/src/engine/canvaskit/image.dart [hunk 2]
```diff
@@ -116,8 +126,14 @@ class CkImage implements ui.Image {
   // being garbage-collected, or by an explicit call to [delete].
   late final SkiaObjectBox box;
 
-  CkImage(this.skImage) {
-    box = SkiaObjectBox(this, skImage as SkDeletable);
+  CkImage(SkImage skImage) : this._(skImage, null);
+
+  CkImage._(this.skImage, SkiaObjectBox? boxToClone) {
+    if (boxToClone != null) {
+      box = boxToClone.clone(this);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/image.dart:133]: same comment as in `CkAnimatedImage`.

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/canvaskit/image.dart (hunk 2 above)
Comment (dnfield) [reply lib/web_ui/lib/src/engine/canvaskit/image.dart:135]: Added these asserts

File: lib/web_ui/lib/src/engine/canvaskit/skia_object_cache.dart [hunk 3]
```diff
@@ -282,6 +282,72 @@ class SkiaObjectBox {
...
+    return null;
+  }
+
+  RefCountedSkiaObjectBox clone(Object wrapper) {
+    assert(!_isDeleted, 'Cannot clone from a deleted handle.');
+    assert(_refs.isNotEmpty);
+    return RefCountedSkiaObjectBox._(wrapper, skObject, _refs);
+  }
+
+  /// Removes the reference count for the [skObject].
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/web_ui/lib/src/engine/canvaskit/skia_object_cache.dart]: Done and done.

**Pull request merged**